bson/src/test/unit/org/bson/codecs/VectorCodecTest.java

```diff
@@ -0,0 +1,162 @@
+                arguments(Vector.packedBitVector(new byte[]{(byte) 0b10101010, (byte) 0b01010101}, (byte) 3), new VectorCodec(), Vector.class)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideVectorsAndCodecsForRoundTrip")
+    void shouldRoundTripVectors(final Vector vectorToEncode) {
+        //given
+        Document expectedDocument = new Document("vector", vectorToEncode);
+
+        //when
+        Codec<Document> codec = CODEC_REGISTRIES.get(Document.class);
+        OutputBuffer buffer = encode(codec, expectedDocument);
+        Document actualDecodedDocument = decode(codec, buffer);
+
+        //then
+        Binary binaryVector = (Binary) actualDecodedDocument.get("vector");
+        assertNotEquals(actualDecodedDocument, expectedDocument);
+        Vector actualVector = binaryVector.asVector();
+        assertEquals(actualVector, vectorToEncode);
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideVectorsAndCodecsForRoundTrip")
+    void shouldEncodeVector(final Vector vectorToEncode, final Codec<Vector> vectorCodec) {
+        // given
+        BsonWriter mockWriter = Mockito.mock(BsonWriter.class);
+
+        // when
+        vectorCodec.encode(mockWriter, vectorToEncode, EncoderContext.builder().build());
+
+        // then
+        verify(mockWriter, times(1)).writeBinaryData(new BsonBinary(vectorToEncode));
+        verifyNoMoreInteractions(mockWriter);
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideVectorsAndCodecsForRoundTrip")
+    void shouldDecodeVector(final Vector vectorToDecode, final Codec<Vector> vectorCodec) {
+        // given
+        BsonReader mockReader = Mockito.mock(BsonReader.class);
+        BsonBinary bsonBinary = new BsonBinary(vectorToDecode);
+        when(mockReader.peekBinarySubType()).thenReturn(BsonBinarySubType.VECTOR.getValue());
+        when(mockReader.readBinaryData()).thenReturn(bsonBinary);
+
+        // when
+        Vector decodedVector = vectorCodec.decode(mockReader, DecoderContext.builder().build());
+
+        // then
+        assertNotNull(decodedVector);
+        assertEquals(vectorToDecode, decodedVector);
+    }
+
+
+    @ParameterizedTest
+    @EnumSource(value = BsonBinarySubType.class, mode = EnumSource.Mode.EXCLUDE, names = {"VECTOR"})
+    void shouldThrowExceptionForInvalidSubType(final BsonBinarySubType subType) {
+        // given
+        BsonReader mockReader = Mockito.mock(BsonReader.class);
+        when(mockReader.peekBinarySubType()).thenReturn(subType.getValue());
+
+        Stream.of(new Float32VectorCodec(), new Int8VectorCodec(), new PackedBitVectorCodec())
+                .forEach(codec -> {
+                    // when & then
+                    BSONException exception = assertThrows(BSONException.class, () ->
+                            codec.decode(mockReader, DecoderContext.builder().build()));
+                    assertEquals("Unexpected BsonBinarySubType", exception.getMessage());
+                });
+    }
```

Comment:
I think we can decode data with invalid subtype without using mocking. Let's try and avoid resorting to mocking when it's not needed.